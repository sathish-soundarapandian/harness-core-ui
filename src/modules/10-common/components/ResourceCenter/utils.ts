/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import moment from 'moment'
import type { UserInfo } from 'services/cd-ng'

export const ON_PREM_RELEASE_NODE_LINK = 'https://ngdocs.harness.io/article/556wy85kbo-harness-on-prem-release-notes'
export const SAAS_RELEASE_NODE_LINK = 'https://ngdocs.harness.io/article/7zkchy5lhj-harness-saa-s-release-notes-2022'

export const HARNESS_SEARCH_LINK = 'https://harness.io/search/'
export const HARNESS_UNIVERISITY_LINK = 'https://university.harness.io/page/job-role-certifications'
export const HARNESS_COMMUNITY_LINK = 'https://community.harness.io/'
export const HARNESS_COMMUNITY_SLACK_LINK =
  'https://join.slack.com/t/harnesscommunity/shared_invite/zt-y4hdqh7p-RVuEQyIl5Hcx4Ck8VCvzBw'
export const HARNESS_DOCS_LINK = 'https://docs.harness.io/'
export const HARNESS_API_DOCS_LINK = 'https://apidocs.harness.io/'
export const HARNESS_TUTORIALS = 'https://developer.harness.io/tutorials/get-started'
export const SITE_STATUS_LINK = 'https://status.harness.io/'

export const timestamp = moment.now()
export const HARNESS_SUPPORT_LINK =
  '/sso.html?action=login&brand_id=114095000394&locale_id=1&return_to=https%3A%2F%2Fsupport.harness.io%2Fhc%2Fen-us%2Frequests&src=zendesk&timestamp=' +
  timestamp

export const WHATS_NEW_LINK = `https://docs.harness.io/article/ueeiji09do-what-s-new`
export const EARLY_ACCESS_LINK = `https://docs.harness.io/article/w4krvu96i3-early-access`

export const openZendeskSupport = (e: React.MouseEvent<Element, MouseEvent>): void => {
  e.stopPropagation()
  e.preventDefault()
  window.open(HARNESS_SUPPORT_LINK)
}

export function getReleaseNodeLink(): string {
  switch (window.deploymentType) {
    case 'COMMUNITY':
    case 'ON_PREM': {
      return ON_PREM_RELEASE_NODE_LINK
    }
    default:
      return SAAS_RELEASE_NODE_LINK
  }
}
export const openFileATicket = (
  e: React.MouseEvent<Element, MouseEvent>,
  currentUserInfo: UserInfo,
  setShow?: React.Dispatch<React.SetStateAction<boolean>>
): void => {
  e.stopPropagation()
  e.preventDefault()
  window.Saber.do?.('set_options', {
    feedback_values: {
      Email: currentUserInfo.email // set default value for email field
    }
  })
  setShow?.(false)
  window.Saber.do?.('open')
}
export const openWhatsNew = (e: React.MouseEvent<Element, MouseEvent>): void => {
  e.stopPropagation()
  e.preventDefault()
  window.open(WHATS_NEW_LINK)
}
export const openEarlyAccess = (e: React.MouseEvent<Element, MouseEvent>): void => {
  e.stopPropagation()
  e.preventDefault()
  window.open(EARLY_ACCESS_LINK)
}
