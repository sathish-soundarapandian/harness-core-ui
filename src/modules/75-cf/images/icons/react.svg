<svg fill="none" height="32" viewBox="0 0 33 32" width="33" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><clipPath id="a"><path d="m.5 0h32v32h-32z"/></clipPath><g clip-path="url(#a)"><path d="m26.8104 10.977c-.3319-.1147-.6759-.2232-1.0301-.3258.0582-.2385.1116-.4739.1592-.70524.7797-3.79915.2699-6.85978-1.4711-7.86746-1.6694-.96624-4.3995.04122-7.1568 2.4497-.2651.23159-.531.4768-.7968.73378-.1771-.17003-.3539-.33433-.5302-.4915-2.8898-2.5753-5.7863-3.66057-7.52553-2.64996-1.66777.96904-2.16167 3.84631-1.45976 7.4468.06779.34777.14702.70288.23661 1.06398-.40993.1168-.80569.2413-1.18436.3738-3.38789 1.1855-5.55156 3.0436-5.55156 4.9709 0 1.9905 2.3227 3.987 5.85146 5.1976.27847.0955.56743.1858.86521.2716-.09666.3905-.18073.7732-.2513 1.1465-.66928 3.538-.14662 6.3473 1.51672 7.3102 1.71811.9944 4.60151-.0277 7.40921-2.4911.2219-.1947.4446-.4012.6677-.6177.2892.2795.5779.5441.8651.7922 2.7196 2.3489 5.4056 3.2975 7.0674 2.3319 1.7164-.9973 2.2742-4.0152 1.55-7.6869-.0553-.2804-.1196-.5669-.1919-.8585.2025-.0601.4013-.1221.5951-.1866 3.6685-1.2199 6.0553-3.1922 6.0553-5.2092 0-1.9343-2.2334-3.8049-5.6896-4.999zm-.7956 8.9058c-.175.0582-.3546.1144-.5375.1688-.405-1.2868-.9516-2.6552-1.6205-4.069.6383-1.38 1.1638-2.7309 1.5574-4.0094.3274.095.6451.1953.9513.3011 2.9612 1.0232 4.7675 2.536 4.7675 3.7017 0 1.2416-1.9508 2.8534-5.1182 3.9068zm-1.3143 2.614c.3202 1.6236.3659 3.0915.1538 4.239-.1906 1.0311-.5739 1.7186-1.0478 1.9939-1.0084.5859-3.165-.1757-5.4909-2.1845-.2666-.2303-.5351-.4762-.8046-.7364.9017-.9898 1.8029-2.1405 2.6824-3.4184 1.5469-.1378 3.0085-.3631 4.3339-.6706.0652.2643.1232.5235.1732.777zm-13.291 6.1317c-.9853.3492-1.77.3592-2.24435.0847-1.00938-.5843-1.42901-2.8399-.85662-5.8656.06556-.3465.1436-.7025.23353-1.0663 1.31087.2909 2.76174.5003 4.31234.6265.8855 1.2504 1.8126 2.3999 2.747 3.4063-.2041.1978-.4075.3861-.6097.5636-1.2415 1.0892-2.4857 1.8621-3.5822 2.2508zm-4.61592-8.7533c-1.56035-.5353-2.84895-1.231-3.73222-1.9901-.79368-.6823-1.19439-1.3596-1.19439-1.9091 0-1.1696 1.73721-2.6613 4.63457-3.6752.35156-.1231.71958-.239 1.10149-.3479.40042 1.3073.92571 2.6742 1.55961 4.0564-.64211 1.4028-1.17485 2.7915-1.5792 4.1141-.27227-.0785-.53605-.1612-.78986-.2482zm1.5473-10.57138c-.60138-3.08481-.20198-5.41187.80308-5.99581 1.07054-.62208 3.43784.26487 5.93294 2.48851.1595.14213.3196.29089.4801.44479-.9297 1.00203-1.8484 2.14281-2.726 3.38592-1.505.14001-2.9456.36492-4.27014.66607-.08327-.33619-.15704-.66653-.21998-.98948zm13.80332 3.42098c-.3166-.5489-.6417-1.0849-.9731-1.6062 1.021.1295 1.9993.3015 2.9179.5119-.2758.8872-.6195 1.8148-1.0242 2.7657-.2902-.5531-.5973-1.1108-.9206-1.6714zm-5.6289-5.5029c.6305.68564 1.262 1.45116 1.883 2.28173-.6258-.02967-1.2594-.04511-1.8983-.04511-.6328 0-1.2618.0151-1.8838.04425.6217-.82285 1.2587-1.58734 1.8991-2.28087zm-5.6651 5.5124c-.3163.5505-.6176 1.1052-.90328 1.6609-.39809-.9476-.73866-1.8794-1.01685-2.7792.91301-.205 1.88663-.3727 2.90113-.4998-.3361.5261-.6638 1.066-.981 1.618zm1.0102 8.1993c-1.0482-.1174-2.0364-.2764-2.94918-.4758.28263-.9158.63071-1.8675 1.03723-2.8356.28645.5553.58895 1.1102.90715 1.6619h.0001c.3242.5621.66 1.1126 1.0047 1.6495zm4.6927 3.8932c-.6479-.7017-1.2942-1.4778-1.9254-2.3127.6128.0241 1.2374.0365 1.8723.0365.6522 0 1.297-.0148 1.9317-.0431-.6232.8499-1.2523 1.6272-1.8786 2.3193zm6.5248-7.2539c.4277.9785.7882 1.9252 1.0745 2.8254-.9278.2125-1.9296.3836-2.985.5107.3322-.5283.6601-1.0735.9823-1.6343.326-.5674.6355-1.1356.9282-1.7018zm-2.1123 1.0163c-.5002.8706-1.0138 1.7016-1.5351 2.486-.9496.0682-1.9306.1033-2.9305.1033-.9958 0-1.9645-.0311-2.8972-.0918-.5424-.7948-1.0671-1.6283-1.5642-2.49h.0002c-.4958-.8594-.952-1.726-1.3654-2.5878.4133-.8638.8683-1.7314 1.3612-2.5893l-.0001.0002c.4942-.8601 1.0143-1.6896 1.5516-2.4793.9517-.0722 1.9276-.1099 2.9138-.1099h.0001c.9907 0 1.9678.038 2.9192.1108.5292.784 1.0458 1.6107 1.5419 2.471.5017.8699.9626 1.7317 1.3795 2.577-.4156.8598-.8755 1.7303-1.375 2.5998zm2.82-15.32201c1.0716.62024 1.4883 3.12168.815 6.40183-.0429.2093-.0913.42248-.1438.63858-1.3276-.3075-2.7693-.5363-4.2786-.67842-.8793-1.25678-1.7905-2.39934-2.7051-3.3885.2459-.23743.4914-.46382.736-.67747 2.3624-2.06357 4.5705-2.87831 5.5765-2.29602zm-7.2856 9.84071c1.5781 0 2.8575 1.284 2.8575 2.8681 0 1.5839-1.2794 2.868-2.8575 2.868s-2.8575-1.2841-2.8575-2.868c0-1.5841 1.2794-2.8681 2.8575-2.8681z" fill="#00d8ff"/></g></svg>