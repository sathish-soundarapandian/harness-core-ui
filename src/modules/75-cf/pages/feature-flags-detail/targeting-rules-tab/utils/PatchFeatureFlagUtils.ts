/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */
import { isEqual } from 'lodash-es'
import type { FeatureState } from 'services/cf'
import patch from '@cf/utils/instructions'
import {
  FormVariationMap,
  VariationTargetGroup,
  TargetingRuleItemType,
  TargetingRulesFormValues,
  VariationPercentageRollout
} from '../types'

// Utils class to help encapsulate the complexity around patch instruction creation and hide this from the components.
interface PatchFeatureFlagUtilsReturn {
  hasFlagStateChanged: () => boolean
  hasDefaultOnVariationChanged: () => boolean
  hasDefaultOffVariationChanged: () => boolean
  addedTargetGroups: (variationIdentifier: string) => VariationTargetGroup[]
  removedTargetGroups: (variationIdentifier: string) => VariationTargetGroup[]
  addedTargets: (variationIdentifier: string) => string[]
  removedTargets: (variationIdentifier: string) => string[]
  updatedPercentageRollouts: () => VariationPercentageRollout[]
  createUpdateFlagStateInstruction: () => void
  createDefaultServeOnInstruction: () => void
  createDefaultServeOffInstruction: () => void
  createAddTargetGroupInstructions: (
    variationIdentifier: string,
    targetGroups: VariationTargetGroup[],
    position: number
  ) => void
  createRemoveTargetGroupsInstructions: (targetGroups: VariationTargetGroup[]) => void
  createAddTargetsInstructions: (variationIdentifier: string, targetIds: string[]) => void
  createRemoveTargetsInstructions: (variationIdentifier: string, targetIds: string[]) => void
  createAddPercentageRolloutInstructions: (percentageRollout: VariationPercentageRollout, index: number) => void
  createUpdatePercentageRolloutInstructions: (percentageRollouts: VariationPercentageRollout[]) => void
  createRemovePercentageRolloutInstruction: (percentageRollout: VariationPercentageRollout) => void
}

export const PatchFeatureFlagUtils = (
  submittedValues: TargetingRulesFormValues,
  initialValues: TargetingRulesFormValues
): PatchFeatureFlagUtilsReturn => {
  const initialPercentageRollouts = initialValues.targetingRuleItems.filter(
    targetingRule => targetingRule.type === TargetingRuleItemType.PERCENTAGE_ROLLOUT
  ) as VariationPercentageRollout[]

  const submittedPercentageRollouts = submittedValues.targetingRuleItems.filter(
    targetingRule => targetingRule.type === TargetingRuleItemType.PERCENTAGE_ROLLOUT
  ) as VariationPercentageRollout[]

  const initialVariations = initialValues.targetingRuleItems.filter(
    targetingRule => targetingRule.type === TargetingRuleItemType.VARIATION
  ) as FormVariationMap[]

  const submittedVariations = submittedValues.targetingRuleItems.filter(
    targetingRule => targetingRule.type === TargetingRuleItemType.VARIATION
  ) as FormVariationMap[]

  const hasFlagStateChanged = (): boolean => submittedValues.state !== initialValues.state

  const hasDefaultOnVariationChanged = (): boolean => submittedValues.onVariation !== initialValues.onVariation

  const hasDefaultOffVariationChanged = (): boolean => submittedValues.offVariation !== initialValues.offVariation

  const addedTargetGroups = (variationIdentifier: string): VariationTargetGroup[] => {
    const initialTargetGroups: VariationTargetGroup[] =
      initialVariations.find(x => x.variationIdentifier === variationIdentifier)?.targetGroups || []
    const submittedTargetGroups: VariationTargetGroup[] =
      submittedVariations.find(x => x.variationIdentifier === variationIdentifier)?.targetGroups || []

    return submittedTargetGroups.filter(
      submittedTargetGroup => !initialTargetGroups.map(({ label }) => label).includes(submittedTargetGroup.label)
    )
  }

  const removedTargetGroups = (variationIdentifier: string): VariationTargetGroup[] => {
    const initialTargetGroups: VariationTargetGroup[] =
      initialVariations.find(x => x.variationIdentifier === variationIdentifier)?.targetGroups || []
    const submittedTargetGroups: VariationTargetGroup[] =
      submittedVariations.find(x => x.variationIdentifier === variationIdentifier)?.targetGroups || []

    return initialTargetGroups.filter(
      targetGroup => !submittedTargetGroups.map(({ label }) => label).includes(targetGroup.label)
    )
  }

  const addedTargets = (variationIdentifier: string): string[] => {
    const initialTargetIds: string[] =
      initialVariations.find(x => x.variationIdentifier === variationIdentifier)?.targets.map(x => x.value) || []
    const submittedTargetIds: string[] =
      submittedVariations.find(x => x.variationIdentifier === variationIdentifier)?.targets.map(x => x.value) || []

    return submittedTargetIds.filter(id => !initialTargetIds.includes(id))
  }

  const removedTargets = (variationIdentifier: string): string[] => {
    const initialTargetIds: string[] =
      initialVariations.find(x => x.variationIdentifier === variationIdentifier)?.targets.map(x => x.value) || []
    const submittedTargetIds: string[] =
      submittedVariations.find(x => x.variationIdentifier === variationIdentifier)?.targets.map(x => x.value) || []

    return initialTargetIds.filter(id => !submittedTargetIds.includes(id))
  }

  const updatedPercentageRollouts = (): VariationPercentageRollout[] => {
    return submittedPercentageRollouts.length === initialPercentageRollouts.length
      ? submittedPercentageRollouts.filter(
          initial => !initialPercentageRollouts.some(submitted => isEqual(initial, submitted))
        )
      : []
  }

  // INSTRUCTIONS SECTION
  const createUpdateFlagStateInstruction = (): void =>
    patch.feature.addInstruction(patch.creators.setFeatureFlagState(submittedValues.state as FeatureState))

  const createDefaultServeOnInstruction = (): void =>
    patch.feature.addInstruction(patch.creators.updateDefaultServeByVariation(submittedValues.onVariation))

  const createDefaultServeOffInstruction = (): void =>
    patch.feature.addInstruction(patch.creators.updateOffVariation(submittedValues.offVariation))

  const createAddTargetGroupInstructions = (
    variationIdentifier: string,
    targetGroups: VariationTargetGroup[],
    priority: number
  ): void => {
    patch.feature.addAllInstructions(
      targetGroups.map((targetGroup, index) =>
        patch.creators.addRule({
          uuid: targetGroup.ruleId,
          priority: priority + index + 1,
          serve: {
            variation: variationIdentifier
          },
          clauses: [
            {
              op: 'segmentMatch',
              values: [targetGroup.value]
            }
          ]
        })
      )
    )
  }

  const createRemoveTargetGroupsInstructions = (targetGroups: VariationTargetGroup[]): void => {
    patch.feature.addAllInstructions(
      targetGroups.map(targetGroup => patch.creators.removeRule(targetGroup.ruleId as string))
    )
  }

  const createAddTargetsInstructions = (variationIdentifier: string, targets: string[]): void => {
    patch.feature.addInstruction(patch.creators.addTargetsToVariationTargetMap(variationIdentifier, targets))
  }

  const createRemoveTargetsInstructions = (variationIdentifier: string, removedTargetIds: string[]): void => {
    patch.feature.addInstruction(
      patch.creators.removeTargetsToVariationTargetMap(variationIdentifier, removedTargetIds)
    )
  }

  const createAddPercentageRolloutInstructions = (
    percentageRollout: VariationPercentageRollout,
    index: number
  ): void => {
    patch.feature.addInstruction(
      patch.creators.addRule({
        uuid: percentageRollout.ruleId,
        priority: index + 1,
        serve: {
          distribution: {
            bucketBy: percentageRollout.bucketBy,
            variations: percentageRollout.variations
          }
        },
        clauses: [
          {
            op: 'segmentMatch',
            values: percentageRollout.clauses[0].values
          }
        ]
      })
    )
  }

  const createUpdatePercentageRolloutInstructions = (percentageRollouts: VariationPercentageRollout[]): void => {
    percentageRollouts.forEach(percentageRollout => {
      const { bucketBy, variations } = percentageRollout

      patch.feature.addInstruction(
        patch.creators.updateRuleVariation(percentageRollout.ruleId as string, { bucketBy, variations })
      )
      const { attribute, negate, op, id, values } = percentageRollout.clauses[0]
      patch.feature.addInstruction(
        patch.creators.updateClause(percentageRollout.ruleId as string, id as string, {
          attribute,
          negate,
          op,
          values
        })
      )
    })
  }

  const createRemovePercentageRolloutInstruction = (percentageRollout: VariationPercentageRollout): void => {
    patch.feature.addInstruction(patch.creators.removeRule(percentageRollout.ruleId as string))
  }

  return {
    hasFlagStateChanged,
    hasDefaultOnVariationChanged,
    hasDefaultOffVariationChanged,
    createUpdateFlagStateInstruction,
    addedTargetGroups,
    removedTargetGroups,
    addedTargets,
    removedTargets,
    updatedPercentageRollouts,
    createDefaultServeOnInstruction,
    createDefaultServeOffInstruction,
    createAddTargetGroupInstructions,
    createRemoveTargetGroupsInstructions,
    createAddTargetsInstructions,
    createRemoveTargetsInstructions,
    createAddPercentageRolloutInstructions,
    createUpdatePercentageRolloutInstructions,
    createRemovePercentageRolloutInstruction
  }
}
