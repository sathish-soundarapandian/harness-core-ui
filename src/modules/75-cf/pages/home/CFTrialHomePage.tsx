/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { useStrings } from 'framework/strings'
import { useFeatureFlag } from '@common/hooks/useFeatureFlag'
import { FeatureFlag } from '@common/featureFlags'
import { CFTrialTemplate } from './CFTrialTemplate'

const CFTrialHomePage: React.FC = () => {
  const { getString } = useStrings()

  const startTrialProps = {
    description: getString('cf.cfTrialHomePage.startTrial.description'),
    learnMore: {
      description: getString('cf.learnMore'),
      url: 'https://docs.harness.io/article/0a2u2ppp8s-getting-started-with-continuous-features'
    },
    startBtn: {
      description: useFeatureFlag(FeatureFlag.FREE_PLAN_ENABLED)
        ? getString('cf.cfTrialHomePage.startFreePlanBtn')
        : getString('cf.cfTrialHomePage.startTrial.startBtn.description')
    }
  }

  return <CFTrialTemplate cfTrialProps={startTrialProps} />
}

export default CFTrialHomePage
