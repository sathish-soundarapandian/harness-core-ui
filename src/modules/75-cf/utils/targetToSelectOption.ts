import type { SelectOption } from '@harness/uicore'
import type { Target } from 'services/cf'

export default function targetToSelectOption({ name, identifier }: Target): SelectOption {
  return { label: name, value: identifier }
}
