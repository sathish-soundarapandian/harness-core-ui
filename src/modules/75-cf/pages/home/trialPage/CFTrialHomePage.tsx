/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { Heading, Layout, Container } from '@harness/uicore'
import { Color, FontVariation } from '@harness/design-system'
import { useParams } from 'react-router-dom'
import { FeatureFlag } from '@common/featureFlags'
import { useFeatureFlag } from '@common/hooks/useFeatureFlag'
import { useStrings } from 'framework/strings'
import {
  ResponseModuleLicenseDTO,
  StartFreeLicenseQueryParams,
  StartTrialDTORequestBody,
  useStartFreeLicense,
  useStartTrialLicense
} from 'services/cd-ng'
import type { AccountPathProps } from '@common/interfaces/RouteInterfaces'
import { Editions } from '@common/constants/SubscriptionTypes'
import { getSavedRefererURL } from '@common/utils/utils'
import bgImageURL from '../ff.svg'
import CFTrialPanel from './CFTrialPanel'
import css from './CFTrialPage.module.scss'

const CFTrialHomePage: React.FC = () => {
  const { getString } = useStrings()
  const { accountId } = useParams<AccountPathProps>()
  const isFreeEnabled = useFeatureFlag(FeatureFlag.FREE_PLAN_ENABLED)
  const refererURL = getSavedRefererURL()

  const startTrialRequestBody: StartTrialDTORequestBody = {
    moduleType: 'CF',
    edition: Editions.ENTERPRISE
  }

  const { mutate: startTrial, loading: startingTrial } = useStartTrialLicense({
    queryParams: {
      accountIdentifier: accountId,
      ...(refererURL ? { referer: refererURL } : {})
    }
  })

  const { mutate: startFreePlan, loading: startingFree } = useStartFreeLicense({
    queryParams: {
      accountIdentifier: accountId,
      moduleType: 'CF' as StartFreeLicenseQueryParams['moduleType']
    },
    requestOptions: {
      headers: {
        'content-type': 'application/json'
      }
    }
  })

  function handleStartTrial(): Promise<ResponseModuleLicenseDTO> {
    return isFreeEnabled ? startFreePlan() : startTrial(startTrialRequestBody)
  }

  const startTrialProps = {
    description: getString('cf.cfTrialHomePage.startTrial.description'),
    learnMore: {
      description: getString('cf.learnMore'),
      url: 'https://docs.harness.io/article/0a2u2ppp8s-getting-started-with-continuous-features'
    },
    startBtn: {
      description: useFeatureFlag(FeatureFlag.FREE_PLAN_ENABLED)
        ? getString('cf.cfTrialHomePage.startFreePlanBtn')
        : getString('cf.cfTrialHomePage.startTrial.startBtn.description')
    }
  }

  return (
    <Container className={css.body} style={{ background: `transparent url(${bgImageURL}) no-repeat` }}>
      <Layout.Vertical spacing="medium">
        <Heading className={css.heading} font={{ variation: FontVariation.H1 }} color={Color.BLACK_100}>
          {getString('cf.cfTrialHomePage.featureFlagsDescription')}
        </Heading>
        <CFTrialPanel {...startTrialProps} startTrial={handleStartTrial} loading={startingTrial || startingFree} />
      </Layout.Vertical>
    </Container>
  )
}

export default CFTrialHomePage
