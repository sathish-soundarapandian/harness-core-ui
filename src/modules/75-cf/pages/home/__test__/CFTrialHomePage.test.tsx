/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render, RenderResult, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { TestWrapper } from '@common/utils/testUtils'
import * as roleAssignmentModal from '@rbac/modals/RoleAssignmentModal/useRoleAssignmentModal'
import * as cdServices from 'services/cd-ng'
import CFTrialHomePage from '../trialPage/CFTrialHomePage'

jest.mock('services/cd-ng', () => ({
  useStartTrialLicense: jest.fn().mockReturnValue({
    cancel: jest.fn(),
    loading: false,
    mutate: jest.fn().mockReturnValue({
      status: 'SUCCESS',
      data: {
        licenseType: 'TRIAL'
      }
    })
  }),
  useStartFreeLicense: jest.fn().mockReturnValue({
    cancel: jest.fn(),
    loading: false,
    mutate: jest.fn().mockReturnValue({
      status: 'SUCCESS',
      data: {
        licenseType: 'FREE'
      }
    })
  })
}))

const renderComponent = (defaultFlagValue?: any): RenderResult =>
  render(
    <TestWrapper
      pathParams={{ accountId: 'dummy', orgIdentifier: 'dummy', projectIdentifier: 'dummy' }}
      defaultFeatureFlagValues={defaultFlagValue}
    >
      <CFTrialHomePage />
    </TestWrapper>
  )

describe('CFTrialHomePage', () => {
  window.deploymentType = 'SAAS'
  const openRoleAssignmentModalMock = jest.fn()

  jest.spyOn(roleAssignmentModal, 'useRoleAssignmentModal').mockReturnValue({
    openRoleAssignmentModal: openRoleAssignmentModalMock
  } as any)

  beforeEach(() => {
    jest.clearAllMocks()
  })

  test('it should show Start Free Plan description if trial is not enabled', async () => {
    window.deploymentType = 'SAAS'
    renderComponent()
    expect(screen.getByText('cf.cfTrialHomePage.startFreePlanBtn')).toBeVisible()
  })

  test('it should show Start Free Plan page with corresponding for Developers descriptions if trial is not enabled', async () => {
    renderComponent()
    expect(screen.getByText('cf.cfTrialHomePage.forDevelopers.title')).toBeVisible()
    expect(screen.getByText('cf.cfTrialHomePage.forDevelopers.createFlag')).toBeVisible()
    expect(screen.getByText('cf.cfTrialHomePage.forDevelopers.shipCode')).toBeVisible()
    expect(screen.getByText('cf.cfTrialHomePage.forDevelopers.realTime')).toBeVisible()
  })

  test('it should show Start Free Plan page with corresponding for DevOps descriptions if trial is not enabled', async () => {
    renderComponent()
    expect(screen.getByText('cf.cfTrialHomePage.forDevOps.title')).toBeVisible()
    expect(screen.getByText('cf.cfTrialHomePage.forDevOps.automatedFeature')).toBeVisible()
    expect(screen.getByText('cf.cfTrialHomePage.forDevOps.avoidRollbacks')).toBeVisible()
    expect(screen.getByText('cf.cfTrialHomePage.forDevOps.scaleManagement')).toBeVisible()
  })

  test('it should try to open the Role Assignment modal when the Invite Developer button is clicked', async () => {
    renderComponent()

    expect(openRoleAssignmentModalMock).not.toHaveBeenCalled()

    await userEvent.click(screen.getByText('cf.cfTrialHomePage.dontCode.inviteDeveloper'))

    await waitFor(() => {
      expect(openRoleAssignmentModalMock).toHaveBeenCalled()
    })
  })

  test('it should start free plan when Start Free Plan is clicked', async () => {
    const handleStartButtonClick = jest.fn()
    window.deploymentType = 'SAAS'
    jest.spyOn(cdServices, 'useStartFreeLicense').mockReturnValue({
      mutate: handleStartButtonClick,
      loading: false
    } as any)

    renderComponent()

    expect(handleStartButtonClick).not.toHaveBeenCalled()

    await userEvent.click(screen.getByText('cf.cfTrialHomePage.startFreePlanBtn'))

    await waitFor(() => expect(handleStartButtonClick).toHaveBeenCalled())
  })

  test('it should start trial when Start 14 day FF Enterprise trial is clicked', async () => {
    const handleEnterpriseButtonClick = jest.fn()
    window.deploymentType = 'ON_PREM'
    jest.spyOn(cdServices, 'useStartTrialLicense').mockReturnValue({
      mutate: handleEnterpriseButtonClick,
      loading: false
    } as any)

    renderComponent()

    expect(handleEnterpriseButtonClick).not.toHaveBeenCalled()
    const text = screen.queryByText('cf.cfTrialHomePage.startTrial.startBtn.description')

    expect(text).not.toBeInTheDocument()
  })
})
