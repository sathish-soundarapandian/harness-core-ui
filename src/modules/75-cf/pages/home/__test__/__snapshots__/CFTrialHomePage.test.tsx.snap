// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CFTrialHomePage snapshot test should render properly 1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main body"
    style="background: transparent url(test-file-stub) no-repeat;"
  >
    <div
      class="Layout--vertical Layout--layout-spacing-medium StyledProps--main"
    >
      <h1
        class="StyledProps--font StyledProps--font-h1 StyledProps--main StyledProps--font-weight-bold StyledProps--font-size-large StyledProps--color StyledProps--color-black100"
      >
        cf.continuous
      </h1>
      <div
        class="Layout--vertical Layout--layout-spacing-small StyledProps--main"
      >
        <p
          class="StyledProps--font StyledProps--main StyledProps--padding-bottom-xxlarge"
          style="width: 500px;"
        >
          cf.cfTrialHomePage.startTrial.description
        </p>
        <a
          class="learnMore"
          href="https://docs.harness.io/article/0a2u2ppp8s-getting-started-with-continuous-features"
          rel="noreferrer"
          target="_blank"
        >
          cf.learnMore
        </a>
        <button
          class="bp3-button Button--button StyledProps--font StyledProps--main StyledProps--intent-primary Button--with-current-color"
          style="width: 300px; height: 45px;"
          type="button"
        >
          <span
            class="bp3-button-text"
          >
            cf.cfTrialHomePage.startTrial.startBtn.description
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
